### PR TITLE
feat(api): create new endpoints

### PR DESCRIPTION
**Pull request by ShaunSHamilton**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Merge me first

_Happy, Oliver?_

### CODE REVIEW

**Review by naomi-lgbt — DISMISSED** [review]
Let's start getting things in.

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:190]:
```suggestion
```
My understanding is that the map will be called when an updatePromise resolves. I'm not sure why you'd want to do this instead of `updatePromise.then(() => res.json...`, though.

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:224]:
```suggestion
    return res.status(400).json({
```
Not sure why these were ever 403s, tbh.  Bad Request seems like a much better fit.

**Comment by ojeytonwilliams** [comment]
There are a few issues mean I think this shouldn't go in just yet, hence the blocked status.  Review's on the way!

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:304]: `getChallenges` isn't desperately expensive (~0.5ms on machine), but we don't need to compute it in response to a request.  It would be better to call it once and use the result.

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:305]: challenge-types is imported, so we don't need magic numbers here.

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:316]: Same argument as above.

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:325]: Same again - no need for magic here.

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:336]: Nit: property not field, right?

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Thanks a ton for creating all the tests. That does, indeed, make me happy.

I added the blocked label because I was worried about multiple `getChallenges` calls.  It's not disastrously slow, so panic over, but there's no need to call in it in a request handler.

File: api-server/src/server/boot/challenge.js
Comment (ShaunSHamilton) [reply api-server/src/server/boot/challenge.js:56]: Well, then we would be looping over the challenges 3 times

**Comment by Sboonny** [comment]
Just wondering, how bad is magic number and how could they have affected the code?

Looking at [magic number wiki](https://en.m.wikipedia.org/wiki/Magic_number_(programming)), I couldn't get what would be the effect on the code.

**Comment by ShaunSHamilton** [comment]
@Sboonny 
> how bad is magic number and how could they have affected the code?

At the end of the day, the code would work exactly the same, _magic_ numbers or no _magic_ numbers. The only issue is, without a way to identify where the numbers come from, the readability/maintainability of the code decreases.

For example, the following code snippets do exactly the same thing, but two are easily understandable, whilst one is unwieldy:

```js
const CURRENT_VAT_RATE = 0.15;
const muhammed_shoe = 10;
const cost = muhammed_shoe + muhammed_shoe * CURRENT_IINTEREST_RATE;
console.log("Muhammed's shoe costs: ", cost);
```

```js
const muhammed_shoe = 10;
// VAT is currently at 15%
const cost = muhammed_shoe + muhammed_shoe * 0.15;
console.log("Muhammed's shoe costs: ", cost);
```

```js
const cost = 10 + 10 * 0.15;
console.log("Muhammed's shoe costs: ", cost);
```

---

If you have more questions on this, I suggest we take the conversation off of this PR, just to keep things neat, and reduce unnecessarily pinging folks.

**Comment by ojeytonwilliams** [comment]
@ShaunSHamilton any more changes in the works or can I give this another review?

**Comment by ShaunSHamilton** [comment]
@ojeytonwilliams This should be ready for another review.

File: utils/challenge-types.js
Comment (lynxlynxlynx) [inline utils/challenge-types.js:46]: Can we resolve this and not have a discussion in the code comments?

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:196]:
```suggestion
```
Yep, we haven't been storing it:
https://github.com/freeCodeCamp/freeCodeCamp/blob/d5423745d963a97fa33dc4584ade8d3e4d126551/api-server/src/server/boot/challenge.js#L284-L288

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:274]:
```suggestion
  const { files, id } = body;
  const completedDate = Date.now();
```
In contrast to the batches, we don't need to trust the client here.

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:364]: These comments are a little confusing and will definitely become obsolete (this won't just be RWD for long).  How about putting them as examples in `challenge-types.js`?

File: api-server/src/server/boot_tests/challenge.test.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot_tests/challenge.test.js:45]: What's the rationale for removing these tests?

File: api-server/src/server/boot/challenge.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/challenge.js:230]: Same issue here: I think we should be getting this from the client.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Aside from a few redundant comments and the missing tests, this is looking good.

I'm not 100% sure what the best approach is for dates, but I suspect we have to trust the client.  Setting all challenges in a batch to have the same `completedDate` doesn't seem right.

File: api-server/src/server/boot_tests/challenge.test.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot_tests/challenge.test.js:335]: Can we just get rid of these?

**Review by ojeytonwilliams — DISMISSED** [review]
This is looking good.

What do you think to adding a single commit with the client changes for QA?  Then we can do some quick checks before dropping that commit and merging this.

File: api-server/src/server/boot/challenge.js
Comment (moT01) [inline api-server/src/server/boot/challenge.js:391]: This seems to break backend and python project submissions - I tried to submit, and it didn't work cause I didn't include a github link. That's always been optional.

**Review by moT01 — COMMENTED** [review]
Found one bug I think. Looking good other than that.

**Comment by raisedadead** [comment]
It's either waiting for review or blocked. All issues and PRs can have only one status at a time.

**Comment by ShaunSHamilton** [comment]
> It's either waiting for review or blocked. All issues and PRs can have only one status at a time.

Right, but until that previous commit is removed, this is blocked and waiting review

**Pull request closed without merging**
